File: api/src/main/java/me/iska/jclient/network/Connection.java
package me.iska.jclient.network;

import me.iska.jclient.network.handler.IHandler;
import me.iska.jclient.network.handler.handlers.LoginHandler;
import me.iska.jclient.network.packet.Packet;
import me.iska.jclient.network.packet.Registry;
import me.iska.jclient.user.User;

import java.io.*;
import java.math.BigInteger;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketTimeoutException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.function.Supplier;
import java.util.logging.Logger;
import java.util.zip.DataFormatException;
import java.util.zip.Deflater;
import java.util.zip.Inflater;

public class Connection extends Thread {

    private static final Logger logger = Logger.getLogger("jclient");

    public static Logger getLogger() {
        return logger;
    }

    private final List<Class<? extends Packet>> capabilities = new ArrayList<>();

    private final List<Packet> latestRecv = new ArrayList<>();
    private final List<Packet> sendQueue = new ArrayList<>();

    private final List<IHandler> secondaryHandlers = new ArrayList<>();

    private final String host;
    private final int port;

    private final ServerInfo serverInfo;

    private Socket sock;
    private InputStream inputStream;
    private OutputStream outputStream;

    private IHandler primaryHandler;

    private boolean connected;
    private long lastPacket;

    private Supplier<Account> accountSupplier;
    private User serverUser;

    private boolean exited;
    private boolean shouldExit;
    private String exitReason;

    public Connection(String host, int port) {
        this.host = host;
        this.port = port;

        capabilities.addAll(Registry.KNOWN_PACKETS);

        serverInfo = new ServerInfo("", false, EncryptionType.NONE, true, 1024, true);

        connected = false;
        lastPacket = System.currentTimeMillis();

        exited = false;
        shouldExit = false;
        exitReason = "Generic disconnect.";
    }

    @Override
    public String toString() {
        return String.format("Connection(host=%s, port=%d)", host, port);
    }

    @Override
    @SuppressWarnings("BusyWait")
    public void run() {
        while (connected) {
            try {
                primaryHandler.update();
                new ArrayList<>(secondaryHandlers).forEach(IHandler::update);
            } catch (Exception error) {
                logger.warning("Error while updating handlers:");
                logger.throwing(Connection.class.getSimpleName(), "run", error);
                exit(error.toString());
            }

            doReadPackets();
            doSendPackets();

            if (System.currentTimeMillis() - lastPacket > 30000) exit("Timed out.");

            if (shouldExit) { // We still want to read and write packets while exiting
                try {
                    primaryHandler.exit(exitReason);
                    new ArrayList<>(secondaryHandlers).forEach(handler -> handler.exit(exitReason));
                } catch (Exception error) {
                    logger.warning("Error while handling exit:");
                    logger.throwing(Connection.class.getSimpleName(), "run", error);
                }

                if (!sendQueue.isEmpty() && connected) { // Don't send if we aren't still connected
                    try {
                        doSendPackets();
                        Thread.sleep(1000);
                    } catch (Exception ignored) {
                    }
                }

                logger.info(String.format("%s disconnected, reason: %s", this, exitReason));

                try {
                    sock.close();
                } catch (IOException error) {
                    logger.warning("Error while closing socket:");
                    logger.throwing(Connection.class.getSimpleName(), "run", error);
                }

                connected = false;
                exited = true;
            }

            /*
            try {
                Thread.sleep(5);
            } catch (InterruptedException ignored) {
            }
             */
        }
    }

    /* ----------------------------- Handlers and connection stuff ----------------------------- */

    public IHandler getPrimaryHandler() {
        return primaryHandler;
    }

    public void setPrimaryHandler(IHandler primaryHandler) {
        this.primaryHandler = primaryHandler;
    }

    public void addSecondaryHandler(IHandler handler) {
        if (!secondaryHandlers.contains(handler)) secondaryHandlers.add(handler);
    }

    public void removeSecondaryHandler(IHandler handler) {
        secondaryHandlers.remove(handler);
    }

    public void setAccountSupplier(Supplier<Account> accountSupplier) {
        this.accountSupplier = accountSupplier;
    }

    public void connect(Socket sock) throws IOException {
        this.sock = sock;
        this.sock.connect(new InetSocketAddress(host, port));

        connected = true;
        lastPacket = System.currentTimeMillis();

        shouldExit = false;
        exitReason = "Generic disconnect.";

        inputStream = sock.getInputStream();
        outputStream = sock.getOutputStream();

        primaryHandler = new LoginHandler(this);
        start();
    }

    /* ----------------------------- Reading packets ----------------------------- */

    private void doReadPackets() {
        int readTimeout = 2; // sendQueue.isEmpty() ? 10 : 1;

        for (int index = 0; index < 10; ++index) {
            Packet packet;
            try {
                packet = readPacket(readTimeout);
            } catch (Exception error) {
                if (!(error instanceof SocketTimeoutException)) {
                    logger.warning("Error while reading packets:");
                    logger.throwing(Connection.class.getSimpleName(), "doReadPackets", error);

                    connected = false; // We can be sure that we aren't connected anymore
                    exit(error.toString());
                }
                return; // We've read all we can right now probably
            }

            if (packet != null) {
                lastPacket = System.currentTimeMillis();

                try {
                    if (primaryHandler != null) primaryHandler.handlePacket(packet);
                    new ArrayList<>(secondaryHandlers).forEach(handler -> handler.handlePacket(packet));
                } catch (Exception error) {
                    logger.warning("Error while handling packet:");
                    logger.throwing(Connection.class.getSimpleName(), "doReadPackets", error);
                    exit(error.toString());
                }
            }
        }
    }

    private Packet readPacket(int readTimeout) throws IOException {
        synchronized (this) {
            byte[] headerBytes = new byte[7];
            int bytesRead = 0;
            sock.setSoTimeout(readTimeout);

            while (bytesRead < 7) {
                int justRead = inputStream.read(headerBytes, bytesRead, 7 - bytesRead);
                sock.setSoTimeout(30000); // Set this immediately so we don't timeout on next read, this mistake caused A LOT of pain
                bytesRead += justRead;
                if (bytesRead == 0) { // We can probably say that the peer hasn't sent us anything
                    return null;
                } else if (justRead < 0) {
                    throw new IOException("Connection closed.");
                }
            }

            ByteArrayInputStream headerInputStream = new ByteArrayInputStream(headerBytes);

            int packetLength = Registry.INTEGER.read(headerInputStream);
            byte[] flags = new byte[1];
            headerInputStream.read(flags);
            int packetID = Registry.UNSIGNED_SHORT.read(headerInputStream);

            bytesRead = 0;

            byte[] packetContents = new byte[packetLength];
            while (bytesRead < packetLength) {
                int justRead = inputStream.read(packetContents, bytesRead, packetLength - bytesRead);
                if (justRead == 0) {
                    return null;
                } else if (justRead < 0) {
                    throw new IOException("Connection closed.");
                }
                bytesRead += justRead;
            }

            if ((flags[0] & 1) == 1) { // Compression
                Inflater inflater = new Inflater();
                ByteArrayOutputStream decompressed = new ByteArrayOutputStream();

                inflater.setInput(packetContents);

                byte[] buffer = new byte[1024];
                while (!inflater.finished()) {
                    int amountRead;
                    try {
                        amountRead = inflater.inflate(buffer);
                    } catch (DataFormatException error) {
                        break;
                    }
                    decompressed.write(buffer, 0, amountRead);
                }

                inflater.end();
                decompressed.close();
                packetContents = decompressed.toByteArray();
            }

            // logger.debug(packetID);

            Optional<Class<? extends Packet>> packetClazz = capabilities.stream()
                    .filter(clazz -> Packet.getID(clazz) == packetID)
                    .findFirst();

            Packet packet = null;

            if (packetClazz.isPresent()) {
                try {
                    packet = packetClazz.get().newInstance();
                } catch (IllegalAccessException | InstantiationException error) {
                    logger.warning(String.format("Couldn't create a new instance of %s:", packetClazz.get()));
                    logger.throwing(Connection.class.getSimpleName(), "readPacket", error);
                    exit("Error while instantiating packet.");
                }
            } else {
                logger.warning(String.format("Unknown packet ID: %d", packetID));
                exit("Received unknown packet.");
            }

            if (packet != null) packet.read(new ByteArrayInputStream(packetContents));

            return packet;
        }
    }

    public Packet getLatestPacket(int timeout) {
        long startedTime = System.currentTimeMillis();
        while (System.currentTimeMillis() - startedTime < timeout) {
            if (!latestRecv.isEmpty()) return latestRecv.remove(0);
        }
        return null;
    }

    /* ----------------------------- Writing packets ----------------------------- */

    private void doSendPackets() {
        for (int index = 0; index < Math.min(15, sendQueue.size()); ++index) {
            try {
                writePacket(sendQueue.remove(0), false);
            } catch (Exception error) {
                if (!(error instanceof SocketTimeoutException)) {
                    logger.warning("Error while sending packets:");
                    logger.throwing(Connection.class.getSimpleName(), "doSendPackets", error);

                    connected = false;
                    exit(error.toString());
                }
                return;
            }
        }
    }

    private void writePacket(Packet packet, boolean noCompression) throws IOException {
        if (packet == null) return;
        // logger.finest(String.format("Writing packet: %s.", packet));

        ByteArrayOutputStream serialized = new ByteArrayOutputStream();
        packet.write(serialized);

        byte flags = 0;

        if (!noCompression && serverInfo.isCompressionEnabled() && serialized.size() > serverInfo.getCompressionThreshold()) {
            flags |= 1;

            Deflater deflater = new Deflater();
            ByteArrayOutputStream compressed = new ByteArrayOutputStream();

            deflater.setInput(serialized.toByteArray());
            deflater.finish();

            byte[] buffer = new byte[1024];
            while (!deflater.finished()) {
                int amountWritten = deflater.deflate(buffer);
                compressed.write(buffer, 0, amountWritten);
            }

            deflater.end();
            compressed.close();

            serialized = new ByteArrayOutputStream();
            serialized.write(compressed.toByteArray());
        }

        Registry.INTEGER.write(serialized.size(), outputStream);
        outputStream.write(flags);
        Registry.UNSIGNED_SHORT.write(Packet.getID(packet.getClass()), outputStream);
        outputStream.write(serialized.toByteArray());
        // outputStream.flush();
    }

    public void sendPacket(Packet packet, boolean instant, boolean noCompression) {
        if (instant) {
            try {
                writePacket(packet, noCompression);
            } catch (IOException error) {
                exit(error.toString());
            }
        } else {
            sendQueue.add(packet);
        }
    }

    public void sendPacket(Packet packet, boolean instant) {
        sendPacket(packet, instant, false);
    }

    public void sendPacket(Packet packet) {
        sendPacket(packet, false);
    }

    /* ----------------------------- Capabilities ----------------------------- */

    public List<Class<? extends Packet>> getCapabilities() {
        return new ArrayList<>(capabilities);
    }

    public void addCapability(Class<? extends Packet> packetClazz) {
        if (!capabilities.contains(packetClazz)) capabilities.add(packetClazz);
    }

    public void setCapabilities(List<Class<? extends Packet>> capabilities) {
        this.capabilities.clear();
        this.capabilities.addAll(capabilities);
    }

    public void addCapabilities(List<Class<? extends Packet>> capabilities) {
        this.capabilities.addAll(capabilities);
    }

    public void removeCapability(Class<? extends Packet> packetClazz) {
        capabilities.remove(packetClazz);
    }

    /* ----------------------------- Other ----------------------------- */

    public void exit(String reason) {
        if (shouldExit) return;

        logger.finer(String.format("Exiting, reason: %s", reason));
        shouldExit = true;
        exitReason = reason;
    }

    public void exit() {
        exit("Generic disconnect.");
    }

    /* ----------------------------- Setters and getters ----------------------------- */

    public String getHost() {
        return host;
    }

    public int getPort() {
        return port;
    }

    public ServerInfo getServerInfo() {
        return serverInfo;
    }

    public InputStream getInputStream() {
        return inputStream;
    }

    public void setInputStream(InputStream inputStream) {
        doReadPackets();
        synchronized (this) {
            this.inputStream = inputStream;
        }
    }

    public OutputStream getOutputStream() {
        return outputStream;
    }

    public void setOutputStream(OutputStream outputStream) {
        doSendPackets();
        synchronized (this) {
            this.outputStream = outputStream;
        }
    }

    public Account getAccount() {
        return accountSupplier == null ? new Account("", "", "") : accountSupplier.get();
    }

    public User getServerUser() {
        return serverUser;
    }

    public void setServerUser(User serverUser) {
        this.serverUser = serverUser;
    }

    public boolean isConnected() {
        return connected;
    }

    public boolean isExited() {
        return exited;
    }

    public String getExitReason() {
        return exitReason;
    }

    /* ----------------------------- Inner classes ----------------------------- */

    /**
     * A representation of an account, for simplicity.
     */
    public static class Account {

        private final String username;
        private final String password;
        private final String groupName;

        public Account(String username, String password, String groupName) {
            this.username = username;
            this.password = password;
            this.groupName = groupName;
        }

        public String getUsername() {
            return username;
        }

        public String getPassword() {
            return password;
        }

        public String getGroupName() {
            return groupName;
        }
    }

    /**
     * Stores information about the server we are connected to.
     */
    public static class ServerInfo {

        private String serverName;

        private boolean encryptionEnabled;
        private EncryptionType encryptionType;

        private boolean compressionEnabled;
        private int compressionThreshold;

        private boolean authenticationEnabled;

        public ServerInfo(String serverName, boolean encryptionEnabled, EncryptionType encryptionType,
                          boolean compressionEnabled, int compressionThreshold, boolean authenticationEnabled) {
            this.serverName = serverName;
            this.encryptionEnabled = encryptionEnabled;
            this.encryptionType = encryptionType;
            this.compressionEnabled = compressionEnabled;
            this.compressionThreshold = compressionThreshold;
            this.authenticationEnabled = authenticationEnabled;
        }

        /**
         * @return The reported name of the server.
         */
        public String getServerName() {
            return serverName;
        }

        public void setServerName(String serverName) {
            this.serverName = serverName;
        }

        /**
         * @return Whether or not encryption is enabled on the server.
         */
        public boolean isEncryptionEnabled() {
            return encryptionEnabled;
        }

        public void setEncryptionEnabled(boolean encryptionEnabled) {
            this.encryptionEnabled = encryptionEnabled;
        }

        /**
         * @return The type of encryption the server wants us to use.
         */
        public EncryptionType getEncryptionType() {
            return encryptionType;
        }

        public void setEncryptionType(EncryptionType encryptionType) {
            this.encryptionType = encryptionType;
        }

        /**
         * @return Whether or not compression is enabled on the server.
         */
        public boolean isCompressionEnabled() {
            return compressionEnabled;
        }

        public void setCompressionEnabled(boolean compressionEnabled) {
            this.compressionEnabled = compressionEnabled;
        }

        /**
         * @return The compression threshold.
         */
        public int getCompressionThreshold() {
            return compressionThreshold;
        }

        public void setCompressionThreshold(int compressionThreshold) {
            this.compressionThreshold = compressionThreshold;
        }

        /**
         * @return Whether or not authentication is enabled on the server.
         */
        public boolean isAuthenticationEnabled() {
            return authenticationEnabled;
        }

        public void setAuthenticationEnabled(boolean authenticationEnabled) {
            this.authenticationEnabled = authenticationEnabled;
        }
    }
}

File: viewer/pyclient/networking/connection.py
#!/usr/bin/env python3

import io
import logging
import socket
import threading
import time
import zlib

from collections import deque
from typing import Tuple

from . import packets
from .handlers import Handler
from .handlers.handshake import HandShakeHandler
from .packets import Packet, DisconnectPacket
from .types.basic import Integer, UnsignedShort
from .types.enum import EncryptionType
from ..impl.user import User


class ServerInfo:
    """
    Stores information about the server.
    """

    @property
    def server_name(self) -> str:
        """
        :return: The reported name of the server.
        """

        return self._server_name

    @server_name.setter
    def server_name(self, server_name: str) -> None:
        self._server_name = server_name

    @property
    def encryption(self) -> bool:
        """
        :return: Whether or not encryption is enabled.
        """

        return self._encryption_enabled

    @encryption.setter
    def encryption(self, encryption_enabled: bool) -> None:
        self._encryption_enabled = encryption_enabled

    @property
    def encryption_type(self) -> EncryptionType:
        """
        :return: The type of encryption the server uses.
        """

        return self._encryption_type

    @encryption_type.setter
    def encryption_type(self, encryption_type: EncryptionType) -> None:
        self._encryption_type = encryption_type

    @property
    def compression(self) -> bool:
        """
        :return: Whether or not compression is enabled.
        """

        return self._compression_enabled

    @compression.setter
    def compression(self, compression_enabled: bool) -> None:
        self._encryption_enabled = compression_enabled

    @property
    def compression_threshold(self) -> int:
        """
        :return: The compression threshold.
        """

        return self._compression_threshold

    @compression_threshold.setter
    def compression_threshold(self, compression_threshold: int) -> None:
        self._compression_threshold = compression_threshold

    @property
    def authentication(self) -> bool:
        """
        :return: Whether or not authentication is enabled.
        """

        return self._authentication_enabled

    @authentication.setter
    def authentication(self, authentication_enabled: bool) -> None:
        self._authentication_enabled = authentication_enabled

    def __init__(self, server_name: str, encryption_enabled: bool, encryption_type: EncryptionType,
                 compression_enabled: bool, compression_threshold: int, authentication_enabled: bool) -> None:
        self._server_name = server_name

        self._encryption_enabled = encryption_enabled
        self._encryption_type = encryption_type

        self._compression_enabled = compression_enabled
        self._compression_threshold = compression_threshold

        self._authentication_enabled = authentication_enabled

    def __repr__(self) -> str:
        return "ServerInfo(name=%r)" % self._server_name


class Connection(threading.Thread):
    """
    A client connection.
    """

    @property
    def host(self) -> str:
        """
        :return: The host address of the server.
        """

        return self._host

    @property
    def port(self) -> int:
        """
        :return: The port of the server.
        """

        return self._port

    @property
    def expecting_timeout(self) -> bool:
        return self._expecting_timeout

    @expecting_timeout.setter
    def expecting_timeout(self, expecting_timeout: bool) -> None:
        if expecting_timeout and not self._expecting_timeout:
            self._last_packet = time.time()

        self._expecting_timeout = expecting_timeout

    @property
    def protocol_version(self) -> int:
        """
        :return: The client's current protocol version.
        """

        return self._protocol_version

    @property
    def connected(self) -> bool:
        """
        :return: Whether or not the client is connected.
        """

        return self._connected

    @property
    def account(self) -> Tuple[str, str, str]:
        """
        :return: The account the client will login with.
        """

        return ("", "", "") if self._account_supplier is None else self._account_supplier()

    @property
    def user(self) -> User:
        """
        :return: The server user associated with the client.
        """

        return self._server_user

    @user.setter
    def user(self, user: User) -> None:
        self._server_user = user

    @property
    def server_info(self) -> ServerInfo:
        """
        :return: The server information.
        """

        return self._server_info

    def __init__(self, host: str, port: int, protocol_version: int = 4,
                 family: int = socket.AF_INET, sock_type: int = socket.SOCK_STREAM):
        super().__init__()

        self._host = host
        self._port = port
        self._protocol_version = protocol_version

        self._expecting_timeout = True

        self._connected = False
        self._exit = False
        self._is_exiting = False
        self.exit_reason = "Generic disconnect."

        self._last_packet = time.time()

        self._sock = socket.socket(family, sock_type)
        self._net_lock = threading.RLock()
        self._read_lock = threading.RLock()

        self._received_packets = deque()
        self._send_queue_packets = deque()

        self.handler = None
        self._other_handlers = []

        self._account_supplier = None

        self._server_info = ServerInfo("", False, EncryptionType.NONE, False, 256, False)
        self._server_user = None

    def __repr__(self) -> str:
        return "Connection(host=%s, port=%i)" % (self._host, self._port)

    def run(self) -> None:
        while not self._exit and not self._is_exiting:
            with self._read_lock:
                if self.handler is not None:
                    try:
                        self.handler.on_update()
                        for handler in self._other_handlers:
                            handler.on_update()
                    except Exception as error:
                        logging.warning("An error occurred during a handler update:")
                        logging.error(error, exc_info=True)
                        self.exit("Exception during handler update.")

            with self._net_lock:
                self._read_packets()
                self._send_packets()

            if self._expecting_timeout and time.time() - self._last_packet > 30:
                self.exit("Timed out.")

            if self._exit:
                self._is_exiting = True

                if self.handler is not None:
                    self.handler.on_exit(self.exit_reason)
                    for handler in self._other_handlers:
                        handler.on_exit(self.exit_reason)

                if len(self._send_queue_packets):
                    self._send_packets()
                    time.sleep(0.1)

                logging.info("Client disconnect, reason: %r." % self.exit_reason)

                self._connected = False
                self._sock.close()

    # -------------------- Misc connection stuff -------------------- #

    def set_account_supplier(self, account_supplier) -> None:
        """
        Set the user and password suppliers for when the server requests auth details.

        :param account_supplier: The supplier for the account details.
        """

        self._account_supplier = account_supplier

    def register_handler(self, handler: Handler) -> None:
        """
        Registers a handler to listen for packets.

        :param handler: The handler to register.
        """

        if not handler in self._other_handlers:
            self._other_handlers.append(handler)

    def unregister_handler(self, handler: Handler) -> None:
        """
        Unregisters a handler from this connection.

        :param handler: The handler to unregister.
        """

        if handler in self._other_handlers:
            self._other_handlers.remove(handler)

    def connect(self) -> None:
        """
        Connects to the server.
        """

        if self._exit:  # Don't connect if we've already exited
            return

        logging.info("Connecting to %s:%i..." % (self._host, self._port))

        self._sock.connect((self._host, self._port))
        self._sock.setblocking(False)

        self._connected = True

        self.handler = HandShakeHandler(self)

        logging.debug("Starting rw thread...")
        self.start()

    # -------------------- Packet read -------------------- #

    def _attempt_read_packet(self, read_timeout: float) -> Packet:
        try:
            self._sock.settimeout(read_timeout)
            header_bytes = self._sock.recv(7)
        except socket.timeout:
            return None

        if not len(header_bytes):
            time.sleep(0.1)  # Peer may have disconnected, let's wait to find out

            if not self._is_exiting:
                raise ConnectionAbortedError("Connection to peer has been aborted (no information).")
            else:
                return None

        self._sock.settimeout(30)

        while len(header_bytes) < 7:
            header_bytes += self._sock.recv(7 - len(header_bytes))

        header_bytes = io.BytesIO(header_bytes)

        packet_length = Integer.read(header_bytes)
        flags = header_bytes.read(1)[0]
        packet_id = UnsignedShort.read(header_bytes)

        packet_buffer = io.BytesIO()
        packet_buffer.write(self._sock.recv(packet_length))

        while packet_buffer.tell() < packet_length:  # Read excess packet data if we weren't able to read it all
            packet_buffer.write(self._sock.recv(packet_length - packet_buffer.tell()))

        for packet in packets.packets:
            if packet.ID == packet_id:

                if flags & 1:  # Packet is compressed
                    new_packet_buffer = io.BytesIO()

                    packet_buffer = zlib.decompress(packet_buffer.getvalue())
                    new_packet_buffer.write(packet_buffer)
                    packet_buffer = new_packet_buffer

                packet_buffer.seek(0)

                read_packet = packet()
                read_packet.read(packet_buffer)

                return read_packet

        return None

    def _read_packets(self) -> None:
        read_timeout = 0.01
        if len(self._send_queue_packets):
            read_timeout = 0.001  # We have more packets to send so no time to wait for new receives

        for index in range(10):
            try:
                packet = self._attempt_read_packet(read_timeout)
            except Exception as error:
                logging.fatal("An exception occurred while attempting to read a packet:")
                logging.error(error, exc_info=True)
                self.exit("Exception while reading packet: %r." % error)
                return

            if packet is None:
                return

            if self.handler is not None:
                try:
                    self.handler.on_packet(packet)
                    for handler in self._other_handlers:
                        handler.on_packet(packet)
                except Exception as error:
                    logging.warning("An error occurred while handling a packet:")
                    logging.error(error, exc_info=True)
                    self.exit("Exception while handling packet.")

            self._last_packet = time.time()
            with self._read_lock:
                self._received_packets.append(packet)

    # -------------------- Packet write -------------------- #

    def _send_packets(self) -> None:
        for index in range(min(15, len(self._send_queue_packets))):
            self._write_packet_instant(self._send_queue_packets.popleft())

    def _write_packet_instant(self, packet: Packet, no_compression: bool = False) -> None:
        packet_data = io.BytesIO()
        buffer = io.BytesIO()

        # print(packet)

        packet.write(packet_data)
        packet_data = packet_data.getvalue()

        flags = 0

        if self._server_info.compression and len(packet_data) > self._server_info.compression_threshold and \
                not no_compression:
            flags |= 1
            packet_data = zlib.compress(packet_data)

        Integer.write(len(packet_data), buffer)
        buffer.write(bytes([flags]))
        UnsignedShort.write(packet.ID, buffer)
        buffer.write(packet_data)

        try:
            self._sock.settimeout(30)
            self._sock.sendall(buffer.getvalue())
        except Exception as error:
            self.exit(repr(error))

    def send_packet(self, packet: Packet, force: bool = True, no_compression: bool = False) -> None:
        with self._net_lock:
            if not force:
                self._send_queue_packets.append(packet)
            else:
                self._write_packet_instant(packet, no_compression=no_compression)

    def flush_packets(self) -> None:
        """
        Flushes all the packets in the send queue.
        """

        # self._read_packets()
        with self._net_lock:
            self._send_packets()

    def exit(self, reason: str = "Generic disconnect.") -> None:
        """
        Disconnects from the peer, given a reason (or not).

        :param reason: The reason for exiting.
        """

        if self._exit:
            return

        self._exit = True
        self.exit_reason = reason

        self.send_packet(DisconnectPacket(reason))
